### ISSUE TITLE
False positive visible private types with super

### ISSUE THREAD

**Issue by thepowersgang** [rust-lang/rust#25532]
Referencing a private parent type with `super` in a public method of a private module generates an error, despite the type not being exposed outside its scope.

```rust
struct Filesystem;

mod dir {
    pub fn new(fs: &super::Filesystem) {}
}

fn main() { }
```
(playpen http://is.gd/7Hhjfw)

```
<anon>:4:21: 5:38 error: private type in exported type signature
<anon>:4     pub fn new(fs: &super::Filesystem) {}
                             ^~~~~~~~~~~~~~~~~
error: aborting due to previous error
playpen: application terminated with error code 101
```

**Comment by alexcrichton** [comment]
This is actually currently working as intended, the definition of "public" is that the target type is marked with `pub` (see the [associated RFC for details](https://github.com/rust-lang/rfcs/blob/master/text/0136-no-privates-in-public.md)), so in this case the `pub` function `new` is referencing the non-`pub` structure `Filesystem`.

**Comment by thepowersgang** [comment]
Hmm... I think I can see why this is intended behaviour, but I'm not sure it's "correct". This snippet does not expose any types that would not be accessible by those that can see the method.

There is a workaround by using the following snippet, but the fact that this workaround works could also be a bug.
```rust
mod dir {
    pub type Filesystem = super::Filesystem;
}
```

**Comment by alexcrichton** [comment]
The problem is that if `dir::new` were reexported at a different publicly-reachable location then the type `Filesystem` would indeed be a private API in a public signature. It's generally quite difficult to determine if something is publicly reexported somewhere else (as it could be a very long chain of reexports), so it's much easier to think about in terms of "is it `pub` or not?"

The workaround you have reexports `Filesystem` as a public type, so that is working as intended.
